### PR TITLE
fix: update min ts version of `jest` & co to 3.8 to allow bumping dependencies

### PR DESCRIPTION
**Pull request by G-Rath**
This is required in order to eventually switch to sourcing types from `jest` directly (#44365), and also to allow us to update our dependency ranges for `@types/jest` to pull in the right version - right now we're pulling in v24 packages, which in addition to being the wrong type means dependency trees tend to have some huge duplication as `@types` comes before `jest` (so package managers tend to put the v24 packages at the top level).

After this is merged, the minimal TS version supported by `@types/jest` & packages that reference `@types/jest` will need to be updated whenever `jest` ups their min TS version.

While this could be pretty breaking for some, we can't really avoid it as TS 3.8 doesn't become the min for DT until 2022 by which time `jest` might have a min version of TS 5 or something, so we'll be back in the exact same spot.

This pain should be once-off since afterwards we'll just be matching the min version as directed by `jest`.

Also: 

* if you're using < jest 26 then you shouldn't be using @types/jest@26 so you wouldn't be affected
* if you're using >= jest 26 then you're currently consuming some jest 24 types, which is wrong anyway, so the act of upgrading the dependencies in theory would cause breakages
* if you can't upgrade typescript then you can stay on the current version of the types you're using - while not ideal, typescript will need to be updated at some point anyway and we can only do so much in DT land

See #45603, #47545, #48426, and maybe a few other PRs & issues.

~Creating as a draft initially as I expect this to go somewhat horribly wrong for the first few CI runs 😬~

### CODE REVIEW

**Comment by G-Rath** [comment]
I just found #43613 where we do the same bumping of TS 🤦😂

File: types/frisby/index.d.ts
Comment (G-Rath) [inline types/frisby/index.d.ts:8]: I don't actually think this is needed - I can't find any evidence of `@types/jest` being used in this file, but will look into it further once I've got CI green.

**Comment by danger-public** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## frisby ([<kbd>unpkg</kbd>](https://unpkg.com/browse/frisby@latest/))
was missing the following properties:
1. delete
## jest-image-snapshot ([<kbd>unpkg</kbd>](https://unpkg.com/browse/jest-image-snapshot@latest/))
was missing the following properties:
1. updateSnapshotState
## jest-json-schema ([<kbd>unpkg</kbd>](https://unpkg.com/browse/jest-json-schema@latest/))
was missing the following properties:
1. matchersWithFormats
## jest-when ([<kbd>unpkg</kbd>](https://unpkg.com/browse/jest-when@latest/))
was missing the following properties:
1. WhenMock
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 68b52e14810df2587af9a101fa3f67b6f399b0ed
</p>

File: types/frisby/index.d.ts
Comment (G-Rath) [reply types/frisby/index.d.ts:8]: It seems that the original reference was to `jasmine`, and was because it was argumenting the global `jasmine` namespace.

However, while v2 of `@types/frisby` removed this argumenting, the reference was kept in because the author wanted to have `@types/jasmine` pulled in because the library used `jasmine`?

I can understand why, and am not against it, but I think it would be better to remove this reference since `frisby.js` doesn't list `jest` as a dependency (not even as a `peerDependency`) nor does it require/import it anywhere, and because people using TypeScript will need to install `@types/jest` anyway as they'll be consuming them directly.

Additionally, the source code of `frisby` still references `jasmine` rather than `jest`, and was originally switched to `jest` as the `jasmine` types were causing problems (#32191).

@cwoodland @johnny4753 thoughts?

File: types/heft-jest/index.d.ts
Comment (octogonz) [inline types/heft-jest/index.d.ts:6]: Heft currently ships with Jest 25. Seems like this change should wait until Heft upgrades its Jest dependency.

Comment (G-Rath) [reply types/heft-jest/index.d.ts:6]: Any idea of when that might be?

I don't think this'll be a problem so long as Heft can support TypeScript 3.8, which by the looks of it is can.

**Review by jwbay — APPROVED** [review]
TS 4.1 beta is out, so the unofficial DT rule of "don't set the minimum to later than two versions ago" seems met here.

File: types/jest/index.d.ts
Comment (ExE-Boss) [inline types/jest/index.d.ts:31]:
```suggestion
// TypeScript Version: 3.8
```

Comment (G-Rath) [reply types/jest/index.d.ts:31]: See, I'd prefer that for consistency, but the readme says [it should have that word](https://github.com/DefinitelyTyped/DefinitelyTyped#i-want-to-use-features-from-typescript-33-or-above) 🤔

**Comment by orta** [comment]
So as a conceptual change, I get where this PR is going (and making it feasible so the module  `jest` be the source of types is a 👍🏻 ) 

I think we need a @SimenB (or other Jest team member) sign-off for it because the TS minimum version change is gonna break a few builds ;)

**Comment by SimenB** [comment]
Sign-off on what? The change itself seems correct to me - it should have happened when `@types/jest` was bumped to v26 as that's the minimal version supported by Jest itself.

Are you asking for some guarantee the TS version will remain stable in the future? If so I can say that Jest 26 will keep 3.8 as minimum (we verify compatibility on CI). No current plans to bump that minimum in the next major (although I really want labeled tuples, I don't think it's worth the downstream breakage).

Regarding breakage of consumers of this package - maybe wait for Jest 27? That's still months away, so might not be worth it, dunno. Your call 👍

**Comment by G-Rath** [comment]
I agree with @SimenB re this should have been done for Jest 26, so if things break they break. `@types/jest` has been pretty stable now anyway and there's always [`patch-package`](https://www.npmjs.com/package/patch-package) 🤷 

ready to merge

**Comment by G-Rath** [comment]
Ready to merge

**Comment by G-Rath** [comment]
@orta @elibarzilay have I broken the bot, or is it just a slow day..? 🤔 

(maybe some sort of status comment could be useful, saying "I've got this queued for publishing"?)

**Comment by G-Rath** [comment]
731 -> 696 packages 🎉

File: types/frisby/index.d.ts
Comment (G-Rath) [reply types/frisby/index.d.ts:8]: I've made #48935 to remove the reference.

**Pull request merged**
